kernel/src/process/signal/poll.rs

```diff
@@ -37,34 +37,29 @@ impl Pollee {
         }
     }
 
-    /// Returns the current events of the pollee given an event mask.
+    /// Returns the current events of the pollee filtered by the given event mask.
     ///
-    /// If no interesting events are polled and a poller is provided, then
-    /// the poller will start monitoring the pollee and receive event
-    /// notification once the pollee gets any interesting events.
+    /// If a poller is provided, the poller will start monitoring the pollee and receive event
+    /// notification when the pollee receives interesting events.
     ///
-    /// This operation is _atomic_ in the sense that either some interesting
-    /// events are returned or the poller is registered (if a poller is provided).
-    pub fn poll(&self, mask: IoEvents, poller: Option<&mut Poller>) -> IoEvents {
+    /// This operation is _atomic_ in the sense that if there are interesting events, either the
+    /// events are returned or the poller is notified.
+    pub fn poll(&self, mask: IoEvents, poller: Option<&mut AnyPoller>) -> IoEvents {
         let mask = mask | IoEvents::ALWAYS_POLL;
 
-        // Fast path: return events immediately
-        let revents = self.events() & mask;
-        if !revents.is_empty() || poller.is_none() {
-            return revents;
-        }
```

Comment:
Why the fast path is removed? It seems the fast path makes sense?

Reply:
There are several reasons why I remove this fast path. I'll list its pros and cons here, and we'll have to decide between the tradeoffs.

### Pros
 1. The semantics of the `poll` method is unclear if we keep this fast path.
    - As a concrete example, it is not possible for the caller to know if the poller is registered by checking the return value of the `poll` method.
       - If the return value contains the events of interest, (i) the poller may not be registered when the fast path is used, meanwhile, (ii) the poller may be registered when the fast path is not used and a race condition occurs later.
 2. Because of the first point, it will not be possible for us to remove the `register_observer` and `unregister_observer` methods directly.
     - This is because we need a method to unconditionally register the observer (i.e. `register_observer`), but the specific semantics cannot be expressed by the `poll` method with its fast path.
 3. Because of the first point, we need to call the `poll` method in the loop of the `wait_events` method. (Otherwise the poller cannot be guaranteed to be registered). This introduces some strangeness in two ways:
    - This leads to redundant registrations. We call the `poll` methods repeatedly, so the poller can be registered multiple times.
    - After implementing https://github.com/asterinas/asterinas/issues/1356, a call to the `poll` method becomes slightly more expensive (as it may involve calculating I/O events). Repeatedly calling the `poll' method in the `wait_events` loop is simply not necessary and should be avoided.
 4. If we're going to make the pollee completely stateless (https://github.com/asterinas/asterinas/issues/1356#issue-2537879395), this fast path will simply become impossible. We will no longer maintain the I/O events in the pollee, so the fast path has to be removed.
    - If we make the pollee conditionally stateless (https://github.com/asterinas/asterinas/issues/1356#issuecomment-2436849331), the fast path can be preserved. But the fast path can only take effect if there are cached I/O events, which seems to make the semantics of the `poll` method even more obscure.

### Cons

Removing a fast path can hurt performance. However, for the specific fast path, I think the damage is minimal.
 - For the epoll file, the poller must be registered regardless of whether the I/O events of interest are present or not. So the original fast path makes no sense.
 - For the `wait_events` method, before we call the `poll` method, we already had another fast path that called the operation (`try_op`) directly. So the original fast path makes no sense if there are no race conditions (but race conditions are rare, so they can be ignored).
https://github.com/asterinas/asterinas/blob/1f612e45f7fccb4caa3901ee26605f828ec64a06/kernel/src/process/signal/poll.rs#L266-L270
 - The original fast path only makes sense for the `poll`/`select` system call and the `do_poll` method. In this scenario, if we see some I/O events of interest, we don't want to keep registering the poller, instead we will return the I/O events directly to the user. However, this scenario occurs at most once during a `do_poll` method, because after that we'll provide `None` as the `poller`.
https://github.com/asterinas/asterinas/blob/1f612e45f7fccb4caa3901ee26605f828ec64a06/kernel/src/syscall/poll.rs#L69-L73

### Conclusion

So in conclusion, I think removing the fast path brings several pros and only one con. For the only con, the damage should also be minimal.

Reply:
Thank you for your explanation!

I understand that you are trying to implement the `poll` method to ensure that the `Poller` is registered, which essentially serves as a superset of the original `register_observer`. The potential performance benefits of the fast path also make sense to me. Therefore, we can remove the fast path in this case.

However, I’m unclear about the points concerning the stateless pollee. We can revisit this topic later, once the design is finalized.

Reply:
Also note that we're doing exactly the same thing as Linux (Linux does not have a fast path for its poll implementation, see [`tcp_poll`](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp.c#L507) for example). So performance shouldn't be an issue as long as we are comparing to Linux).